engine/lib/phases/phase_rewrite_control_flow.ml

```diff
@@ -34,143 +34,75 @@ module Make (F : Features.T) =
-                            else_ = Some (rewrite_branch else_);
-                          };
-                    }
-                | rhs -> { e with e = Let { monadic = None; lhs; rhs; body } })
-            | Let
-                {
-                  monadic = None;
-                  lhs;
-                  rhs = { e = Match { scrutinee; arms }; _ };
-                  body;
-                } ->
-                let rewrite =
-                  List.fold arms ~init:false ~f:(fun acc (arm : arm) ->
-                      acc || has_return#visit_arm () arm)
+                  U.make_lets
+                    (branch_stmts @ ((p, branch_final) :: stmts_after))
+                    final
                 in
-                if rewrite then
-                  {
-                    e with
-                    e =
-                      Match
-                        {
-                          scrutinee = self#visit_expr () scrutinee;
-                          arms =
-                            List.map arms ~f:(fun arm ->
-                                let arm_body = arm.arm.body in
-                                let arm_body =
-                                  {
-                                    arm_body with
-                                    e =
-                                      Let
-                                        {
-                                          monadic = None;
-                                          lhs;
-                                          rhs = arm_body;
-                                          body;
-                                        };
-                                  }
-                                in
-                                self#visit_arm ()
-                                  {
-                                    arm with
-                                    arm = { arm.arm with body = arm_body };
-                                  });
-                        };
-                  }
-                else e
-            | _ -> super#visit_expr () e
-        end
-
-      (* This visitor allows to remove instructions after a `return` so that `drop_needless_return` can simplify them. *)
-      let remove_after_return =
-        object (self)
-          inherit [_] Visitors.map as super
-
-          method! visit_expr () e =
-            match e.e with
-            | Let { monadic = None; lhs; rhs; body } -> (
-                let rhs = self#visit_expr () rhs in
-                let body = self#visit_expr () body in
-                match rhs.e with
-                | Return _ -> rhs
-                | _ -> { e with e = Let { monadic = None; lhs; rhs; body } })
+                match
+                  List.findi stmts ~f:(fun _ (_, e) ->
+                      match e.e with
+                      | (If _ | Match _) when has_return#visit_expr () e -> true
+                      | Return _ -> true
+                      | _ -> false)
```

Comment:
I think here you can use `split_while`, so that you don't manipulate indexes explicitly and you get rid of `before_after`.